### ISSUE TITLE
[Highlight Proposal] Create new TSNamespace highlight group for 'namespace' language concept

### ISSUE THREAD

**Issue by bluz71** [nvim-treesitter/nvim-treesitter#516]
Hello Neovim Treesitter team,

At the moment C++ namespaces such as `std` are highlighted as `TSConstant`. With my [moonfly](https://github.com/bluz71/vim-moonfly-colors) and [nightfly](https://github.com/bluz71/vim-nightfly-guicolors) themes the results are not ideal.

Recently we had an issue where HTML tags where being highlighted as `TSType` which also was not ideal. New `TSTag` and `TSTagDelimiter` highlight groups resulted from that discussion. This was a big improvement. I believe we have another similar issue here.

Namespaces in programming languages provide a scope to partition logic and functionality under a chosen name to avoid collisions. The most well-known example would be C++, here is an example:

```cpp
using std::cout;
using std::endl;

namespace box1 {
    int box_side = 4;
}

namespace box2 {
    int box_side = 12;
}

int main() {
    int box_side = 42;
    cout << box1::box_side << endl;  // Outputs 4.
    cout << box2::box_side << endl;  // Outputs 12.
    cout << box_side << endl;  // Outputs 42.
}
```

`std`, `box1` and `box2` are namespaces. Ideally in Treesitter I would like to highlight those in a custom color, not as `TSConstant`. In my opinion namespaces are not constants.

Namespaces are not limited to C++.

Ruby has an equivalent mechanism with `module` as seen in this snippet:

```ruby
require 'digest'

module Encryption
  def encrypt(string)
    Digest::SHA2.hexdigest(string)
  end
end
```

`Encryption` and `Digest` are effectively namespaces.

I am not a Java person, but it may be the case the Java packages could be considered namespaces.

C# has namespaces:

```c#
using System;

Console.WriteLine("Hello World!");
int i = Convert.ToInt32("123");
```

`System` is considered a namespace.

Python modules might also be considered namespaces:

```python
import modulea

modulea.func1()
modulea.func2()
a = modulea.class1()
```

`modulea` may be considered a namespace?

Lua may also be a candidate for possible `TSNamespace` highlighting:

```lua
local M = {}

local function test() print(123) end
function M.test1() test() end
function M.test2() M.test1(); M.test1() end

return M
```

`M` may be considered a namespace.

I guess this issue is a request for comment much like `TSTag` was. Feedback welcome.

My suggestion, create a new `TSNamespace` highlight group and associated `@namespace` directive, by default link `TSNamespace` to `Type` or `Constant`, and then rollout to the programming languages that support the concept. That is all predicated on Treesitter itself providing the right tooling to retrieve namespaces to begin with?

Ideally Treesitter should not have an explosion of highlight groups names, however if there are concepts that apply across many languages then it may be better to create specialized highlight groups rather than shoehorning in generic highlight groups.

Best regards.

P.S. I know I have created many support issues recently (serial pest), but this one is near the tail-end of my list (maybe even the last major one). Overall Treesitter highlight quality is really impressive, except for namespaces in C++ (hence this issue).

**Comment by theHamsta** [comment]
This is when you're getting old: where is the namespace highlight? It was always there... I remember seeing it yesterday...

-> No, it is only used for locals...

You are right namespaces deserve their own highlight and not constants as they are right now.

**Comment by vigoux** [comment]
Question : can names paces be used for modules too in some languages? (think of rust atm)

**Comment by theHamsta** [comment]
I'd say yes. That's our policy for locals right now.

**Comment by bluz71** [comment]
> Question : can names paces be used for modules too in some languages? (think of rust atm)

Yes absolutely, that is the intention. Deploy it to all languages that support the concept of namespacing, which is quite a few.

As mentioned in the top post, modules in Ruby are a namespace. Python modules may be as well. Java packages.

This is definitely not a C++ only request, though that is where this started.

**Comment by theHamsta** [comment]
C++ and sometimes Rust are though one of the very few examples where we can distinguish namespaces from normal objects without semantic analysis.

**Comment by bluz71** [comment]
What about Ruby, it uses the `module` keyword for definition and double `::` when calling. 

It should not need any smarts, it is quite similar to C++.

Should semantic analysis be avoided, extra complexity was not the aim of this issue.

**Comment by theHamsta** [comment]
We have extra semantic highlighting as a long term goal #81. We can make it work right now but it is too slow at the moment.

**Comment by bluz71** [comment]
Commits [8a994d2](https://github.com/theHamsta/nvim-treesitter/commit/8a994d25ee7334f4ba80bbda3486ae8342b66466) and [45397eb](https://github.com/nvim-treesitter/nvim-treesitter/commit/45397ebbd9fb2bb13b14824e917158d2e02f960a) have been merged.

Many thanks to the implementers.

But, [this C++ file](https://github.com/bluz71/vim-nightfly-guicolors/blob/master/samples/cplusplus.cc) appears not be be highlighting all namespaces. The `v8::` statements have no color:

![cplusplus](https://user-images.githubusercontent.com/11382509/95668654-b00af080-0bc2-11eb-9572-3709d7f58032.png)

My expectation is they should be `TSNamespace` highlighted. Note, `std::` in this same sample file is being highlighted correctly with the new highlight group.

I also now notice that `XmlNode` on lines 65 and 67 no longer are being highlighted as types:

![cplusplus2](https://user-images.githubusercontent.com/11382509/95668676-01b37b00-0bc3-11eb-9e42-e58835fd093d.png)

Likely this one is a separate issue, but raising here first in case I am missing something obvious.

Best regards.

**Comment by theHamsta** [comment]
The v8 probably only requires to add numbers in the namespace rebex. I thought I included some fixes for the parameters in the "fix method"... I have to look were those changes went

**Comment by bluz71** [comment]
Very strange. I think the images I posted above were using [vim-polyglot](https://github.com/sheerun/vim-polyglot) for C++.

I did install treesitter and it was in effect for Rust, Ruby, Go etc. I don't know what happened with C++. Anyway today treesitter works with C++.

Must have been a solar flare induced false positive :confused:

C++ namespace highlighting now looks great via treesitter:

![treesitter](https://user-images.githubusercontent.com/11382509/95696226-0856f680-0c86-11eb-941f-d84fb0b971f8.png)

Many thanks @theHamsta.

Feature request implemented, this issue issue is now resolved.

P.S. In the image above, C++ methods are not highlighted correctly, still waiting for PR #453 to resolve #446
